### ISSUE TITLE
Service Locator 디자인 패턴 구현

### ISSUE THREAD

**Issue by gyuwon** [envicase/ReactiveMvvm#28]
Service Locator 패턴은 몇 가지 단점으로 인해 많은 공격을 받는 의존성 역전 기법이지만 생성자 주입만으로는 해결하기 어려운 몇 가지 상황에 요긴하게 사용될 수 있습니다. Reactive-MVVM은 기본적인 Service Locator 구현체를 제공하고 코드 품질을 저해하지 않는 절제된 범위에서 사용합니다.

**Comment by jwChung** [comment]
아 이 부분 벌써 진행 중이시군요.
말씀하신 "생성자 주입만으로는 해결하기 어려운 몇 가지 상황"에 팩토리 또는 퍼사드 패턴으로는 해결되지 못할까요?

**Comment by gyuwon** [comment]
@jwChung 댓글을 이제야 확인했네요. 팩토리도 퍼사드도 inject 되거나 locate 되어야 서비스될 수 있지 않을까요? 예를 들어 `ReactiveCommand<T>` 클래스는 주 스레드 스케줄러를 사용합니다. 이것이 생성자 주입을 사용하지 않는 상황에서 팩토리나 퍼사드로 어떻게 해결될 수 있을까요?

**Comment by gyuwon** [comment]
@jwChung 상황이 조금 달라졌습니다. `ReactiveCommand<T>`의 직접적인 스케줄러(`IScheduler`) 지원 기능을 보류하게 되었습니다. 그러다보니 Service Locator와 관련된 요구가 당장은 없게 되었네요. 불필요한 코드가 있을 필요가 없다고 판단되어 우선은 Service Locator와 관련된 코드를 제거하겠습니다. 그렇지만 그것과는 별개로 팩토리나 퍼사드와 관련된 논의는 좀 더 이어갔으면 합니다.

**Comment by jwChung** [comment]
@gyuwon 

1. 생성자 주입을 사용하지 못하는 이유가 무엇인지요?
   제가 생각했던 것은 `Func<Action<object>, Func<object, bool>, ICommand>` 같은 팩토리를 뷰모델로 생성자로 주입하거, 아니면 커멘트 객체 생성을 좀 더 추상화한 퍼사드를 주입하는 것이 어떨까였습니다.
 
2. 아래와 같은 접근도 불가능한가요?
	```c#
	public ICommand FooCommand =>
	    new Command(
	        x => {},
	        x => true,
	        new SomeScheduler(SynchronizationContext.Current));
	```

3. 위 1, 2번 둘다 사실 좋은 코드는 아닌 것 같습니다.
   1번의 경우는 팩토리가 주입될 수 있다하더라도, 예를들어 하나의 커멘드 객체를 만들려고  팩토리를 주입하는 것도 문제고,
   2 번이 된다하더라도 new 생성자로 구체클래스를 생성하고 있으니 이것도 문제네요.
   문제의 근원은 커멘드 객체를 뷰모델에서 생성하고 있는게 아닐까 합니다. 그보다 뷰모델로 조립된 커멘드 객체를 주입하는 것이 문제 해결책 같은데요. 아래와 같이 `[BindingCommand]` 같은 특성을 이용하여 이름 컨벤션(Can커멘드명, 커멘드명Async) 또는 또 다른 특성을 이용하여 `SaveCommand`와 메소드(`CanSave`, `SaveAsync`)를 bind해주면 어떨까 하는 생각해봤습니다. 그래서 ViewModel이 생성되면 기본적으로 `[BindingCommand]`특성을 검사하여 `ICommand` 객체를 Setter 주입하는 것으로요.
	```c#
	public class ViewModel
	{
		[BindingCommand]
		public ICommand SaveCommand{ get; set;}
	
		public bool CanSave(int parameter)
		{
			return true;
		}
	
		public Task SaveAsync(int parameter)
		{
			return Task.FromResult<object>(null);
		}
	}
	```

**Comment by gyuwon** [comment]
@jwChung 1, 2번은 좋지 않다고 말씀하신 그대로가 제 생각이고요, 참고로 3번은 몇 년전 [교보문고의 e서재](http://ebook3.dkyobobook.co.kr/Kyobo_T3/HelpDesk/HelpDesk_eLibrary.asp) 프로젝트에서 제가 만든 프레임워크에서 사용한 방식입니다. 보여주신 방식도 작동했고 더 생략한 문법을 지원하기도 했습니다. MVVM 패턴을 사용하면서 늘어나는 코드가 아주 짜증났었거든요.

```csharp
// SomeViewModel creates 'SaveCommand' command automatically and dynamically.
public class SomeViewModel : ViewModel
{
    public bool CanSave(object parameter)
    {
        return true;
    }

    public void Save(object parameter)
    {
    }
}
```

히지만 이런 방식을 사용하더라고 개별 서비스는 어디선가는 제공되어야 합니다. 이 때 지난 번에 말씀드린 것처럼 상위 수준의 뷰모델이 하위 수준의 뷰모델(상위 수준의 뷰모델이 생성하는)이 사용하는 서비스에 대해 알아야 하느냐는 겁니다. 지난 번에는 피드 뷰모델과 항목 뷰모델, 그리고 `IClipboard` 서비스가 그 예였죠.

```csharp
public class ItemViewModel
{
    private readonly IClipboard _clipboard;

    // Yes, I need a clipboard service.
    public ItemViewModel(IClipboard clipboard, ItemModel model)
    {
        _clipboard = clipboard;
    }

    public void ExecuteCopyUriCommand(object parameter)
    {
        _clipboard.Copy(GetUri());
    }
}

public class FeedViewModel
{
    private readonly IClipboard _clipboard;

    // Why should I know about IClipboard?
    public FeedViewModel(IClipboard clipboard)
    {
        _clipboard = clipboard;
    }

    private void OnItemsLoaded(IEnumerable<ItemModel> models)
    {
        var viewModels = models.Select(m => new ItemViewModel(_clipboard, m)).ToList();
    }
}
```

이 문제를 팩토리나 퍼사드를 사용해서 해결하는 것도 바람직해 보이지는 않습니다. 프로그램이 너무 복잡해지거든요. 테스트는 좀 쉬워질런지 모르겠습니다. 하지만 프로덕션는 이해하기 어려워지고 유지보수성도 낮아지지 않을까요? `FeedViewModel`에 `ItemViewModel`에 대한 팩토리 함수가 주입되었다고 가정합니다. 아직까지는 적당한 수준이라고 예상됩니다. 그런데 둘 사이와 유사한 관계가 `ItemViewModel`과 또 다른 뷰모델 사이에 있다고 생각해보면 코드를 유지보수해야 할 때에 팩토리 함수의 정체를 파악하는 데에만 큰 비용이 들어갈겁니다. 함수가 아닌 인터페이스를 사용하면 코드 추적이 좀 쉬워질 수는 있겠지만 타입이 너무 늘어나는 문제가 발생하죠.

**Comment by jwChung** [comment]
글쎄요 제가 @gyuwon 님이 말씀하시는 것을 잘 이해 못하고 있는건지 모르겠네요. 말씀하신 대로 `ICommand`와 `IClipboard`는 다른 얘기네요.

아래 코드를 이렇게 생각하신다는 말씀이신거죠?

>이 문제를 팩토리나 퍼사드를 사용해서 해결하는 것도 바람직해 보이지는 않습니다. 프로그램이 너무 복잡해지거든요. 테스트는 좀 쉬워질런지 모르겠습니다. 하지만 프로덕션는 이해하기 어려워지고 유지보수성도 낮아지지 않을까요? 

```c#
public class ItemViewModel
{
    private readonly IClipboard _clipboard;

    // Yes, I need a clipboard service.
    public ItemViewModel(IClipboard clipboard, ItemModel model)
    {
        _clipboard = clipboard;
    }

    public void ExecuteCopyUriCommand(object parameter)
    {
        _clipboard.Copy(GetUri());
    }
}

public class FeedViewModel
{
    private readonly IClipboard _viewModelFactory;

    public FeedViewModel(Func<ItemModel, ItemVeiwModel> viewModelFactory)
    {
        _viewModelFactory = viewModelFactory;
    }

    private void OnItemsLoaded(IEnumerable<ItemModel> models)
    {
        var viewModels = models.Select(viewModelFactory.Invoke).ToList();
    }
}

public static void Main(string[] args)
{
    IClipboard clipboard = new ConcreteClipboard();
     var feed = new FeedViewModel(m => new ItemViewModel(clipboard, m));
}
```

**Comment by gyuwon** [comment]
@jwChung 네 맞아요. 보여주신 코드 정도는 나쁘지 않다고 봅니다. 여기에 `ItemViewModel`에도 팩토리 함수가 주입되어야하는 상황이 발생하면(적지 않게 발생할거라고 생각합니다.) 코드가 복잡해지기 시작할거라는 것이죠.

**Comment by jwChung** [comment]
@gyuwon new 객체 생성해서 주입하는 것을 한 곳에(프로그램 시작점) 몰아 넣는 것이 중요할 것 같아요. 

```c#
public static void Main(string[] args)
{
      IClipboard clipboard = new ConcreteClipboard();
      var feed = new FeedViewModel(m => 
          new ItemViewModel(
               clipboard,
               m,
               () => new Foo()));
}
```

**Comment by gyuwon** [comment]
@jwChung 자주 발생할 거라고 예상되는 또 다른 경우는 `ViewModel_A'가 'ViewModel_B`와 `ViewModel_C`를 만드는 경우입니다. 예를 들어 위에서 말씀 드린 e서재의 경우 다양한 종류의 컨텐트(오디오, 비디오, 플래시, ...)를 처리하고 사용자가 목록에서 특정 컨텐트를 선택하면 각 컨텐트 유형을 처리하는 뷰모델을 만들었어야 했습니다. 이런 경우를 팩토리 함수 주입으로 처리하는 것이 바람직할까요?

**Comment by jwChung** [comment]
@gyuwon: Abstract 팩토리 패턴과 같이 묶어서 넣어주면 문제가 될까요? Abstract 팩토리 추상화가 필요할지 없을지는 상황에 따라 달라질 것 같아요.

```c# 
public class MediumFactory
{
     public MediumFactory(IClipboard clipboard){...}
     public Audio CreateAudio(){...}
     public Video CreateVideo(){....}
     public Flash CreateFlash(){....}
}
```

**Comment by gyuwon** [comment]
@jwChung 그런데 이런 류의 팩토리가 조금만 커지면 Service Locator 인스턴스를 주입하는 것과 유사해지는 것 아닌가요? 사실 `IServiceProvider` 인스턴스를 생성자 주입하는 경우도 종종 볼 수 있죠.

**Comment by jwChung** [comment]
@gyuwon 

무엇이 사용되는지 알고 모르고 차이가 아닐까 합니다. 
SL의 문제점으로 얘기되는 뭐가 사용되는지 명시적이지 않다점 말이죠.

`new Factory<T>().Create(); `
`new ServiceLocator().Create<TAny>(); `

아 그렇군요. 뭔길을 돌아온 느낌인네요 결국 SL vs DI 군요. 저를 이해시키느라 수고하셨습니다. :disappointed: ... 

이 상황에서 @gyuwon 님은 SL을 택하겠다는 입장이시구요.

**Comment by gyuwon** [comment]
@jwChung 맞아요. 명시적이지 않은 것이 문제죠. 그런데 팩토리가 커지면 그것 역시 모호함을 가지게 될 거라는 것이 제 생각이에요. 그래서 흑마술이라도 적당히 쓰는 것이 어떻겠냐는 거에요. 아직 Service Locator가 바로 필요한 단계가 되지는 않았으니 다른 방법에 대해서 더 생각해보고 싶긴합니다. 저도 Service Locator에 긍정적인 입장은 아니니까요.

**Comment by testcode77** [comment]
코드를 잘 살펴봤습니다. 저는 익스트림한 성능을 추구하는 경향이 있어서 이 정도면 어느 정도 오버헤드가 있을지 한번 더 올려보고 이야기 드릴 예정입니다만, 뷰가 복잡해지면 복잡해질수록 자원의 효율적인 관리는 이슈가 될 것 같습니다.

이 정도로 통합적인 모델보다 저 같은 경우는 네이티브로 짜서 성능을 향상 시키는 쪽을 선호합니다. 회사 개발 팀 자체의 인적 구성이나 현재 상황에서 이런 모델을 나쁘지 않다고 보긴합니다.

**Comment by gyuwon** [comment]
@testcode77 의견 감사합니다. 말씀하신 것처럼 한정된 자원 활용이란 측면이 강하게 작용한 아키텍처입니다. 간단한 규칙만 지켜서 상태 동기화를 얻어내는 것이 목적입니다. 아직 성능을 심도있게 고려하진 못하고 있지만 앞으로 최대한 튜닝해보고 싶습니다.
